### PR TITLE
feat(store): implement metadata storage

### LINKED ISSUE #110: Implement metadata
We should get a general Metadata layer on the Store, with capabilities to Query on the metadata values.
We can also make TTL a specific Metadata field.

### PR DESCRIPTION
**Pull request by runcom**
Fixes #110

### CODE REVIEW

**Comment by puiterwijk** [comment]
Couple of comments provided via chat about the types.

File: http-wrapper/src/server.rs [hunk 1]
```diff
@@ -20,12 +20,23 @@ pub struct SessionWithStore {
 
 type SessionStoreT = Arc<SessionStore>;
 
+#[non_exhaustive]
+pub enum SessionStoreMetadataKey {}
```
Comment (runcom) [inline http-wrapper/src/server.rs:24]: doesn't need pub

File: manufacturing-server/src/main.rs [hunk 2]
```diff
@@ -37,11 +37,28 @@ struct DiunConfiguration {
     public_keys: PublicKey,
 }
 
+#[non_exhaustive]
+pub enum PublicKeyStoreMetadataKey {}
```
Comment (runcom) [inline manufacturing-server/src/main.rs:41]: doesn't need pub

File: rendezvous-server/src/main.rs [hunk 3]
```diff
@@ -48,10 +48,19 @@ impl Serializable for StoredItem {
     }
 }
 
+#[non_exhaustive]
+pub enum RendezvousStoreMetadataKey {}
```
Comment (runcom) [inline rendezvous-server/src/main.rs:52]: doesn't need pub

File: rendezvous-server/src/handlers_to0.rs [hunk 4]
```diff
@@ -203,6 +202,17 @@ pub(super) async fn ownersign(
         .await
         .map_err(Error::from_error::<messages::to0::OwnerSign, _>)?;
 
+    user_data
+        .store
+        .store_metadata(
+            &device_guid,
+            &fdo_store::MetadataKey::Ttl,
+            &chrono::Duration::from_std(Duration::from_secs(600))
```
Comment (puiterwijk) [inline rendezvous-server/src/handlers_to0.rs:210]: This Ttl is actually not just 600, but depends on the protocol value :).

File: http-wrapper/src/server.rs [hunk 5]
```diff
@@ -39,10 +50,15 @@ impl SessionStore {
...
-                session.id().to_string(),
-                Some(Duration::from_secs(600)),
-                session.clone(),
+            .store_data(session.id().to_string(), session.clone())
+            .await?;
+        self.store
+            .store_metadata(
+                &session.id().to_string(),
+                &fdo_store::MetadataKey::Ttl,
+                &chrono::Duration::from_std(Duration::from_secs(600)).map_err(|_| {
```
Comment (puiterwijk) [inline http-wrapper/src/server.rs:59]: Let's maybe make a global constant for session ttl, so it's simple to see and change?

File: rendezvous-server/src/handlers_to0.rs (hunk 4 above)
Comment (runcom) [reply rendezvous-server/src/handlers_to0.rs:210]: Damn, copy pasted , will fix it I see there’s a ttl variable indeed

File: store/src/directory.rs [hunk 6]
```diff
@@ -82,22 +86,129 @@ fn ttl_from_disk(ttl: &[u8]) -> Result<SystemTime, StoreError> {
...
+            let val = V::deserialize_from_reader(&file);
+            if let Ok(v) = val {
+                values.push(v)
+            } else {
+                log::trace!("Error deserializing data {:?}", r);
+            }
+        }
+        Ok(Some(ValueIter {
+            index: 0,
+            values: values,
```
Comment (puiterwijk) [inline store/src/directory.rs:197]: This can be just `values,`. Rust is nice that way.

Comment (runcom) [reply store/src/directory.rs]: yup just changed

**Review by puiterwijk — APPROVED** [review]

**Pull request merged**